### DISCUSSION TITLE
How to disable SSG crawler

### DISCUSSION THREAD

**Discussion by wildyeast** [nuxt/nuxt#16120] [Questions] [answered]
I use Nuxt 3 for static site generation, which unfortunately seems to have lost important features compared to v2 (see [this discussion](https://github.com/nuxt/framework/discussions/9588)). This is the relevant code in `nuxt.config.ts`:

```
const getSlugs = async () => {
  const baseURL = 'mydomain.com'
  const endpoint = `${baseURL}/api/getSlugs`
  const response = await fetch(endpoint)
  return response.json()
}

export default defineNuxtConfig({
 hooks: {
   async 'nitro:config' (nitroConfig) {
     const slugs = await getSlugs()
     for (const slug of slugs) {
       nitroConfig.prerender?.routes?.push(slug.route)
     }
   }
 },
...
```

When I run `nuxt generate`, this line is printed:

```
ℹ Prerendering 35 initial routes with crawler
```

The prerendered routes are not only those I defined in the previous code snippet, but also all links found on those pages by a crawler. I would like to disable that crawler. In Nuxt 2, you could do that with

```
generate: {
  crawler: false
}
```

but that leads to errors now. How do I do that in the nitro:config hook?

**Comment by hacknug** [comment, marked as answer]
Not sure why you want to use `nitro:config` but the way I did it was adding this to my config file:

```ts
export default defineNuxtConfig({
  generate: {
    crawler: false,
    fallback: true,
    routes: ['/', '/tv', '/movie', '/search'],
  },
}
```

I would try using `generate.routes` with your `getSlugs()`method while disabling the crawler at the same time. That should hopefully result in what you expect unless I'm missing something.

Hope this helps.

**Comment by wildyeast** [comment]
Thank you for your reply! I tried this, but I can only get it to work with an array of route literals, like in your example. Do you know how I would use the array returned by my `getSlugs()` method? I tried
```
 routes: getSlugs()
```
and 
```
 routes: await getSlugs()
```
but it doesn't work. In Nuxt 2 it was possible to pass a function to `routes`:

```javascript
 generate: {
    fallback: true,
    async routes () {
    ...
```
This now throws a TypeScript error:
```
error TS2322: Type '() => Promise<unknown>' is not assignable to type 'any[]'.
```
When I suppress the TypeScript error with `@ts-ignore`, `generate` seems to treat the function code as a string:
```
  ├─ /page_in_pages_folder_0 (5ms)         
  ├─ /page_in_pages_folder_1 (9ms)          
  ├─ /page_in_pages_folder_2 (10ms)         
  ├─ async routes() {                                                                                                                                                                                 nitro 20:18:48
      const slugs = await getSlugs();
      console.log(slugs);
      return slugs;
    } (undefinedms) (TypeError [ERR_INVALID_URL]: Invalid URL)
  ├─ /200.html (2ms)                                                                                                                                                                                  nitro 20:18:48
  ├─ /404.html (0ms)
```

**Reply by hacknug** [reply]
From the types documentation you should be able to use any of this approaches:

```
(property) routes?: any[] | undefined

The routes to generate. If you are using the crawler, this will be only the starting point for route generation. This is often necessary when using dynamic routes. It can be an array or a function.

@example

routes: ['/users/1', '/users/2', '/users/3']

You can pass a function that returns a promise or a function that takes a callback. It should return an array of strings or objects with route and (optional) payload keys.

@example

export default {
  generate: {
    async routes() {
      const res = await axios.get('https://my-api/users')
      return res.data.map(user => ({ route: '/users/' + user.id, payload: user }))
    }
  }
}

Or instead:

export default {
  generate: {
    routes(callback) {
      axios
        .get('https://my-api/users')
        .then(res => {
          const routes = res.data.map(user => '/users/' + user.id)
          callback(null, routes)
        })
        .catch(callback)
    }
  }
}

If routes() returns a payload, it can be accessed from the Nuxt context.

@example

export default {
  async useAsyncData ({ params, error, payload }) {
    if (payload) return { user: payload }
    else return { user: await backend.fetchUser(params.id) }
  }
}
```

If the `async` approach isn't working for you I would try to call your method outside of `defineNuxtConfig()` and pass-in the static value returned by your method. This might solve your issue (it is something I use on some of my Nuxt v2 sites that need to SSG pages for routes defined on an external CMS).

Another thing I'd check is the version of Nuxt you're using. Perhaps it is not the stable one? In case it is, have you tried with the edge version? Maybe there was a bug and it has been fixed there?

**Comment by wildyeast** [comment]
My ideal solution would indeed be this, from your example:

```
generate: {
  async routes() {
    const res = await getSlugs()
    return res.map(slug => ({ route: slug.route , payload: slug }))
  }
}
```
However, this results in the following TS errors:
```
nuxt.config.ts:47:11 - error TS2322: Type '() => Promise<any>' is not assignable to type 'any[]'.

47     async routes() {
             ~~~~~~

nuxt.config.ts:50:14 - error TS18046: 'res' is of type 'unknown'.

50       return res.map(slug => ({ route: slug.route , payload: slug }))
                ~~~

nuxt.config.ts:50:22 - error TS7006: Parameter 'slug' implicitly has an 'any' type.

50       return res.map(slug => ({ route: slug.route , payload: slug }))
                        ~~~~
```
If I disable TS, Nuxt assumes that the function code is a page:

```
  ├─ async routes() {    
      const res = await getSlugs();
      return res.map((slug) => ({ route: slug.route, payload: slug }));
    } (undefinedms) (TypeError: route.split is not a function)
```
This leads me to believe that the documentation you quoted, especially the feature to "pass a function that returns a promise or a function that takes a callback" to `generate.routes`, is not currently implemented. I tried with both v3.0.0 and the edge version with no difference.
